### ISSUE TITLE
[ext_authz] CheckRequest call fails when using multiple http filters

### ISSUE THREAD

**Issue by emanueledomingo** [envoyproxy/envoy#27653]
*Title*: *CheckRequest call fails when using multiple http filters (http_service + grpc_service)*

*Description*:

I have two http filters:

- filter1 is a custom http api that returns 200 or 403.
- filter2 is an external grpc service (open-policy-agent)

The request after being successfully evaluated from the first http filter should be forewarded to the second one.

Some additional notes:

- In another configuration i use the same stack with an OAuth2 filter followed by the external grpc service (OPA) and it works fine. This is an experiment to replace the [Oauth2 Filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/oauth2_filter) with a basic authentication filter.
- The logs of filter2 are empty, i guess the request doesn't even reach the grpc upstream bacause blocked by Envoy
- It seems an error with the CheckRequest call, but no further information given. Just `CheckRequest call failed with status: Internal`.

*Config*:

```yml
static_resources:
  listeners:
    - name: listener
      ...

      filter_chains:
        - filters:
            - name: envoy.filters.network.http_connection_manager
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
                stat_prefix: ingress_http

                http_filters:
                  - name: envoy.filters.http.filter1
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
                      transport_api_version: V3
                      stat_prefix: filter1
                      http_service:
                        server_uri:
                          uri: <address>
                          cluster: filter1_cls
                          timeout: 0.5s

                  - name: envoy.filters.http.filter2
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
                      transport_api_version: V3
                      failure_mode_allow: false
                      grpc_service:
                        google_grpc:
                          target_uri: <address>
                        timeout: 0.5s

                  - name: envoy.filters.http.router
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router

                route_config:
                  ...

  clusters:
    ...

admin:
  ...
```

*Logs*:

```
[2023-05-24 08:50:42.097][16][trace][http] [source/common/http/http1/codec_impl.cc:568] [C197] message begin
[2023-05-24 08:50:42.097][16][trace][http] [source/common/http/http1/codec_impl.cc:525] [C197] completed header: XXX
[2023-05-24 08:50:42.097][16][trace][http] [source/common/http/http1/codec_impl.cc:525] [C197] completed header: XXX
[2023-05-24 08:50:42.097][16][trace][http] [source/common/http/http1/codec_impl.cc:525] [C197] completed header: XXX
[2023-05-24 08:50:42.097][16][trace][http] [source/common/http/http1/codec_impl.cc:525] [C197] completed header: XXX
[2023-05-24 08:50:42.097][16][trace][http] [source/common/http/http1/codec_impl.cc:525] [C197] completed header: XXX
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:525] [C197] completed header: XXX
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:525] [C197] completed header: XXX
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:818] [C197] onHeadersCompleteBase
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:525] [C197] completed header: key=set-cookie value=XXX; HttpOnly; Path=/; SameSite=lax; Secure
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:1411] [C197] status_code 200
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:1421] [C197] Client: onHeadersComplete size=8
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:922] [C197] message complete
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:1492] [C197] message complete
[2023-05-24 08:50:42.098][16][debug][client] [source/common/http/codec_client.cc:126] [C197] response complete
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=1)
[2023-05-24 08:50:42.098][16][trace][router] [source/common/router/upstream_request.cc:221] [C0][S4613335283428937577] upstream response headers:
':status', '200'
'date', 'Wed, 24 May 2023 08:50:40 GMT'
'server', 'uvicorn'
'content-length', '0'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
[2023-05-24 08:50:42.098][16][debug][router] [source/common/router/router.cc:1371] [C0][S4613335283428937577] upstream headers complete: end_stream=true
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=2)
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=3)
[2023-05-24 08:50:42.098][16][debug][http] [source/common/http/async_client_impl.cc:105] async http request response headers (end_stream=true):
':status', '200'
'date', 'Wed, 24 May 2023 08:50:40 GMT'
'server', 'uvicorn'
'content-length', '0'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'set-cookie', 'XXX; HttpOnly; Path=/; SameSite=lax; Secure'
'x-envoy-upstream-service-time', '298'
[2023-05-24 08:50:42.098][16][trace][ext_authz] [source/extensions/filters/http/ext_authz/ext_authz.cc:257] [C196][S8163095975218825222] ext_authz filter added header(s) to the request:
[2023-05-24 08:50:42.098][16][trace][ext_authz] [source/extensions/filters/http/ext_authz/ext_authz.cc:287] [C196][S8163095975218825222] ext_authz filter removed header(s) from the request:
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/filter_manager.cc:69] [C196][S8163095975218825222] continuing filter chain: filter=0x14c27f952fc0
[2023-05-24 08:50:42.098][16][trace][ext_authz] [source/extensions/filters/http/ext_authz/ext_authz.cc:73] [C196][S8163095975218825222] ext_authz filter calling authorization server
[2023-05-24 08:50:42.098][16][trace][ext_authz] [source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:42] Sending CheckRequest: attributes {XXX}
[2023-05-24 08:50:42.098][18][trace][grpc] [source/common/grpc/google_async_client_impl.cc:58] completionThread CQ event 0 false
[2023-05-24 08:50:42.098][16][trace][grpc] [source/common/grpc/google_async_client_impl.cc:230] Queued message to write (4727 bytes)
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/filter_manager.cc:537] [C196][S8163095975218825222] decode headers called: filter=envoy.filters.http.filter2 status=4
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=4)
[2023-05-24 08:50:42.098][16][debug][pool] [source/common/http/http1/conn_pool.cc:53] [C197] response complete
[2023-05-24 08:50:42.098][16][debug][pool] [source/common/conn_pool/conn_pool_base.cc:214] [C197] destroying stream: 0 remaining
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/http1/codec_impl.cc:676] [C197] parsed 4397 bytes
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:125] clearing deferred deletion list (size=4)
[2023-05-24 08:50:42.098][16][trace][grpc] [source/common/grpc/google_async_client_impl.cc:298] handleOpCompletion op=0 ok=false inflight=1
[2023-05-24 08:50:42.098][16][debug][grpc] [source/common/grpc/google_async_client_impl.cc:221] notifyRemoteClose 13 
[2023-05-24 08:50:42.098][16][trace][ext_authz] [source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:116] CheckRequest call failed with status: Internal
[2023-05-24 08:50:42.098][16][trace][ext_authz] [source/extensions/filters/http/ext_authz/ext_authz.cc:417] [C196][S8163095975218825222] ext_authz filter rejected the request with an error. Response status code: 403
[2023-05-24 08:50:42.098][16][debug][http] [source/common/http/filter_manager.cc:905] [C196][S8163095975218825222] Sending local reply with details ext_authz_error
[2023-05-24 08:50:42.098][16][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x14c27f273780 for 300000ms, min is 300000ms
[2023-05-24 08:50:42.098][16][trace][ext_authz] [source/extensions/filters/http/ext_authz/ext_authz.cc:175] [C196][S8163095975218825222] ext_authz filter has 0 response header(s) to add and 0 response header(s) to set to the encoded response:
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/filter_manager.cc:1065] [C196][S8163095975218825222] encode headers called: filter=envoy.filters.http.filter2 status=0
[2023-05-24 08:50:42.098][16][trace][ext_authz] [source/extensions/filters/http/ext_authz/ext_authz.cc:175] [C196][S8163095975218825222] ext_authz filter has 0 response header(s) to add and 0 response header(s) to set to the encoded response:
[2023-05-24 08:50:42.098][16][trace][http] [source/common/http/filter_manager.cc:1065] [C196][S8163095975218825222] encode headers called: filter=envoy.api_auth status=0
[2023-05-24 08:50:42.098][16][debug][http] [source/common/http/conn_manager_impl.cc:1559] [C196][S8163095975218825222] encoding headers via codec (end_stream=true):
':status', '403'
'date', 'Wed, 24 May 2023 08:50:41 GMT'
'server', 'envoy'
[2023-05-24 08:50:42.098][16][trace][connection] [source/common/network/connection_impl.cc:478] [C196] writing 97 bytes, end_stream false
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=1)
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=2)
[2023-05-24 08:50:42.098][16][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x14c27f2f8e00 for 3600000ms, min is 3600000ms
[2023-05-24 08:50:42.098][16][debug][grpc] [source/common/grpc/google_async_client_impl.cc:242] resetStream
[2023-05-24 08:50:42.098][18][trace][grpc] [source/common/grpc/google_async_client_impl.cc:58] completionThread CQ event 5 true
[2023-05-24 08:50:42.098][16][debug][grpc] [source/common/grpc/google_async_client_impl.cc:415] Stream cleanup with 1 in-flight tags
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:125] clearing deferred deletion list (size=2)
[2023-05-24 08:50:42.098][16][trace][grpc] [source/common/grpc/google_async_client_impl.cc:298] handleOpCompletion op=5 ok=true inflight=1
[2023-05-24 08:50:42.098][16][debug][grpc] [source/common/grpc/google_async_client_impl.cc:404] Deferred delete
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=1)
[2023-05-24 08:50:42.098][16][trace][main] [source/common/event/dispatcher_impl.cc:125] clearing deferred deletion list (size=1)
[2023-05-24 08:50:42.099][16][debug][grpc] [source/common/grpc/google_async_client_impl.cc:167] GoogleAsyncStreamImpl destruct
[2023-05-24 08:50:42.099][16][trace][connection] [source/common/network/connection_impl.cc:563] [C196] socket event: 2
[2023-05-24 08:50:42.099][16][trace][connection] [source/common/network/connection_impl.cc:674] [C196] write ready
[2023-05-24 08:50:42.099][16][trace][connection] [source/common/network/raw_buffer_socket.cc:67] [C196] write returns: 97
```

**Comment by wbpcode** [comment]
cc @tyxia
